File: ex_5_opti.py
from collections import defaultdict

class OptimalPageReplacement:
    def __init__(self, frame_count):
        self.frame_count = frame_count
        self.frames = []
        self.page_table = defaultdict(list)

    def page_fault_occurred(self, pages):
        page_faults = 0
        for index, page in enumerate(pages):
            if page not in self.frames:
                page_faults += 1
                if len(self.frames) < self.frame_count:
                    self.frames.append(page)
                else:
                    # Find the page that will not be used for the longest period in the future
                    page_to_replace = max(self.frames, key=lambda x: self.page_table[x][self.page_table[x].index(index)+1:] if index in self.page_table[x] else float('inf'))
                    self.frames[self.frames.index(page_to_replace)] = page
            self.page_table[page].append(index)
        return page_faults

    def print_frames(self):
        print("Current frames:", self.frames)

# Example usage:
if __name__ == "__main__":
    opt = OptimalPageReplacement(3)
    pages = [1, 2, 3, 4, 1, 2, 5, 1, 2, 3, 4]
    page_faults = opt.page_fault_occurred(pages)
    opt.print_frames()
    print("Total number of page faults:", page_faults)

File: ex_5_opti.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <limits>

using namespace std;

class OptimalPageReplacement {
private:
    int frame_count;
    vector<int> frames;
    unordered_map<int, vector<int>> page_table;

public:
    OptimalPageReplacement(int count) : frame_count(count) {}

    bool page_fault_occurred(int page) {
        if (find(frames.begin(), frames.end(), page) == frames.end()) {
            if (frames.size() < frame_count) {
                frames.push_back(page);
            } else {
                int farthest_index = -1;
                int page_to_replace = -1;

                for (int i = 0; i < frames.size(); ++i) {
                    if (page_table[frames[i]].empty()) {
                        page_to_replace = frames[i];
                        break;
                    } else {
                        auto it = find(page_table[frames[i]].begin(), page_table[frames[i]].end(), page);
                        if (it == page_table[frames[i]].end()) {
                            page_to_replace = frames[i];
                            break;
                        } else {
                            int distance = std::distance(page_table[frames[i]].begin(), it);
                            if (distance > farthest_index) {
                                farthest_index = distance;
                                page_to_replace = frames[i];
                            }
                        }
                    }
                }

                replace_page(page, page_to_replace); 
            }

            page_table[page].push_back(frames.size() - 1);
            return true; // Page fault occurred
        }
        return false; // No page fault occurred
    }

    void replace_page(int new_page, int old_page) {
        auto it = find(frames.begin(), frames.end(), old_page);
        *it = new_page;
    }

    void print_frames() {
        cout << "Current frames: ";
        for (int frame : frames) {
            cout << frame << " ";
        }
        cout << endl;
    }
};

int main() {
    OptimalPageReplacement optimal(3);
    vector<int> pages = {1, 2, 3, 4, 1, 2, 5, 1, 2, 3, 4};
    int page_faults = 0;

    for (int page : pages) {
        if (optimal.page_fault_occurred(page)) {
            page_faults++;
        }
        optimal.print_frames();
    }

    cout << "Total number of page faults: " << page_faults << endl;

    return 0;
}
